# Global genome analyses reveal a vast and dynamic anellovirus landscape within the human virome
#### Cesar Arze <sup>1</sup>, Simeon Springer <sup>1</sup>, Gytis Dudas <sup>2</sup>, Sneha Patel <sup>1</sup>, Agamoni Bhattacharya <sup>1</sup>, Harish Swaminathan <sup>1</sup>, Carlo Brugnara <sup>3,4</sup>, Simon Delagrave <sup>1</sup>, Avak Kahvejian <sup>1,5</sup>, Yann Echelard <sup>1,5</sup>, Erica G. Weinstein <sup>1,5</sup>, Roger J. Hajjar <sup>1,5</sup>, Kristian G. Andersen <sup>6</sup>, Nathan L. Yozwiak <sup>1</sup>
<sup>1</sup>Ring Therapeutics, Cambridge, MA, USA,
<sup>2</sup>Gothenburg Global Biodiversity Centre, Gothenburg, Sweden,
<sup>3</sup>Boston Children’s Hospital, Boston, MA, USA,
<sup>4</sup>Harvard Medical School, Boston, MA, USA,
<sup>5</sup>Flagship Pioneering, Cambridge, MA, USA,
<sup>6</sup>Scripps Research Institute, La Jolla, CA, USA,

## Abstract
Viruses inhabit all humans, yet the virome remains understudied. Anelloviruses are a ubiquitous component of the healthy human virome and remain highly prevalent after infecting early in life. It is unclear how vast and diverse the ‘anellome’ is, however, and its evolutionary dynamic over time and space remains unexplored. Here we use in-depth sequencing of blood transfusion donor(s)-recipient pairs coupled with public genomic resources for large-scale assembly of novel anellovirus genomes to characterize global and personal anellovirus diversity through time. We observe that the breadth of the anellome is much greater than previously believed and we show that individuals harbor and transmit a multitude of unique anellomes that can persist for several months within a diverse milieu of endemic host lineages. We show that much of the anellovirus diversity is due to extensive recombination, which, if unaccounted for, hinders phylogenetic analyses. These findings illuminate the transmission dynamics and vast diversity of anelloviruses within the human virome and may offer a novel conduit for therapeutic gene delivery.

## Install
Analysis scripts associated with this manuscript require the following software:

1. Python
   - jupyterlab
   - matplotlib
   - biopython
   - scipy
   - shapely
   - baltic
2. R
   - ggplot2
   - dplyr2
   - reshape2
   - ggthemes
   - tidyverse
   - magrittr
   - here
   - foreach
   - scales
   - ggpubr
   - gridExtra
3. [ClonalFrameML](https://github.com/xavierdidelot/ClonalFrameML)
4. [PhyML](https://github.com/stephaneguindon/phyml)

A conda environment has been provided for ease-of-use setup to run all analysis.

## File Descriptions
__TODO__: Provide information on all data files or any auxillary files required for analysis.

## Sequence Data Availability
Sequence data can be found with the following NCBI's BioProject accession: [PRJNA679286](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA679286)

## Citation
__TODO__: Add citation when available
