{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-16T08:04:44.569861Z",
     "start_time": "2021-04-16T08:04:43.626655Z"
    }
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "\n",
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "\n",
    "import glob,os,pathlib\n",
    "\n",
    "from itertools import islice\n",
    "\n",
    "def window(seq, n=2):\n",
    "    \"Returns a sliding window (of width n) over data from the iterable\"\n",
    "    \"   s -> (s0,s1,...s[n-1]), (s1,s2,...,sn), ...                   \"\n",
    "    it = iter(seq)\n",
    "    result = tuple(islice(it, n))\n",
    "    if len(result) == n:\n",
    "        yield result\n",
    "    for elem in it:\n",
    "        result = result[1:] + (elem,)\n",
    "        yield result\n",
    "\n",
    "typeface='Arial'\n",
    "mpl.rcParams['axes.labelweight']=300\n",
    "mpl.rcParams['font.family']=typeface\n",
    "mpl.rcParams['font.size']=22\n",
    "mpl.rcParams['pdf.fonttype']=42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-16T08:04:44.862374Z",
     "start_time": "2021-04-16T08:04:44.573206Z"
    }
   },
   "outputs": [],
   "source": [
    "base_path=pathlib.Path(os.path.abspath('')).parent\n",
    "data_path=os.path.join(base_path, \"data\", \"figure_1_2C\")\n",
    "\n",
    "seqs={}\n",
    "seq_names={}\n",
    "\n",
    "\n",
    "for aln_path in glob.glob(os.path.join(data_path,'*.fasta')):\n",
    "    print(aln_path)\n",
    "    aln=SeqIO.parse(aln_path,format='fasta')\n",
    "    aln_name=os.path.basename(aln_path).split('_')[0]\n",
    "    \n",
    "    seqs[aln_name]=[]\n",
    "    seq_names[aln_name]=[]\n",
    "    \n",
    "    seq_names[os.path.basename(aln_path)]=[]\n",
    "    for seq in aln:\n",
    "        seq_names[aln_name].append(seq.id)\n",
    "        seqs[aln_name].append(seq.seq)\n",
    "        \n",
    "for aln_path in glob.glob(os.path.join(base_path,'data','figure_S1','*_aa.fasta')):\n",
    "    print(aln_path)\n",
    "    aln=SeqIO.parse(aln_path,format='fasta')\n",
    "    aln_name=os.path.basename(aln_path).split('_')[0]\n",
    "    \n",
    "    seqs[aln_name]=[]\n",
    "    seq_names[aln_name]=[]\n",
    "    \n",
    "    seq_names[os.path.basename(aln_path)]=[]\n",
    "    for seq in aln:\n",
    "        seq_names[aln_name].append(seq.id)\n",
    "        seqs[aln_name].append(seq.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-16T08:04:55.275028Z",
     "start_time": "2021-04-16T08:04:44.880873Z"
    }
   },
   "outputs": [],
   "source": [
    "def entropy(column,alphabet_size=20):\n",
    "    column=Counter(column)\n",
    "    \n",
    "    for invalid in ['-','X','*']:\n",
    "        if invalid in column:\n",
    "            column.pop(invalid)\n",
    "    \n",
    "    total=sum(column.values())\n",
    "    \n",
    "    entropy=0\n",
    "    for x in column:\n",
    "        entropy+=column[x]/total*np.log2(column[x]/total)/np.log2(alphabet_size)\n",
    "\n",
    "    return -entropy\n",
    "\n",
    "def unique_residues(column):\n",
    "    column=Counter(column)\n",
    "    \n",
    "    for invalid in ['-','X','*']:\n",
    "        if invalid in column:\n",
    "            column.pop(invalid)\n",
    "        \n",
    "    return len(column)\n",
    "\n",
    "count_valid=lambda k,cutoff: len(list(filter(lambda w: w not in ['X','-','*'],k)))>cutoff ## remove invalid residues from column, check if the number of remaining residues is above cut-off\n",
    "\n",
    "ys={}\n",
    "\n",
    "for aln in seqs:\n",
    "    print(aln)\n",
    "    ys[aln]=[unique_residues(c) for c in zip(*seqs[aln]) if count_valid(c,len(c)*0.1)] ## return number of unique residues at columns with at least 10% valid (non-gap, non-stop, non-ambiguous) residues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-16T08:05:17.353374Z",
     "start_time": "2021-04-16T08:04:55.311363Z"
    }
   },
   "outputs": [],
   "source": [
    "colours={'Betas': '#3F877A', 'Gammas': '#C8786F', 'Alphas': '#D9D936', 'Anellos': '#6F6B8D'}\n",
    "\n",
    "rename={'IAVgroup2': 'Influenza A virus group 2','HIV1':'HIV-1','AAV':'Adeno-associated virus', 'Alphas': 'Alphatorquevirus', 'Betas': 'Betatorquevirus', 'Gammas': 'Gammatorquevirus', \n",
    "        'Filo': 'Filovirus', 'Lassa': 'Lassa fever virus', 'HPV': 'Human papillomavirus', \n",
    "        'MERS': 'MERS-CoV', 'Dengue': 'Dengue virus'}\n",
    "\n",
    "genes={'Betas': 'ORF1', 'Gammas': 'ORF1', 'Alphas': 'ORF1', 'Anellos': 'ORF1', \n",
    "       'AAV': 'capsid', 'IAVgroup2': 'HA', 'HIV1': 'env', \n",
    "       'Filo': 'GP', 'HPV': 'L1', 'Lassa': 'GP', \n",
    "       'MERS': 'S', 'Dengue': 'E'}\n",
    "\n",
    "\n",
    "fig = plt.figure(figsize=(40, 25),facecolor='w') \n",
    "gs = gridspec.GridSpec(4,3,wspace=0.02,hspace=0.1) \n",
    "\n",
    "\n",
    "for a,aln in enumerate(['Alphas','Betas','Gammas','HIV1','IAVgroup2','AAV','Filo','HPV','Lassa','MERS','Dengue']): ## iterate over proteins\n",
    "    row=a%4\n",
    "    column=a//4\n",
    "    \n",
    "    ax=plt.subplot(gs[row,column],facecolor='w')\n",
    "    \n",
    "    w_size=50 ## window size for sliding window\n",
    "    \n",
    "    fc=colours[aln] if aln in colours else '#949494' ## Anello colour or grey\n",
    "    \n",
    "    ax.plot(range(len(ys[aln])),ys[aln],color=fc,zorder=1) ## plot raw data\n",
    "\n",
    "    pys=list(map(np.mean,window(ys[aln],w_size))) ## mean sliding window across raw data\n",
    "    \n",
    "    ax.plot(range(int(w_size/2),int(len(pys)+w_size/2)),pys,lw=3,color='k',zorder=10) ## plot sliding window in black\n",
    "\n",
    "    print('%20s \\tmean: %.3f\\tst.dev.: %.3f'%(aln,np.mean(ys[aln]),np.std(ys[aln])))\n",
    "    \n",
    "    w=3 ## width of arrow\n",
    "    arrow_props={'head_length': 10,\n",
    "                 'width': w,\n",
    "                 'head_width': w, \n",
    "                 'length_includes_head': True, \n",
    "                 'edgecolor': 'none', \n",
    "                 'facecolor': fc} ## arrow parameters\n",
    "    \n",
    "    y=21.05\n",
    "    ax.arrow(0,y+arrow_props['width']/2,len(ys[aln]),0,**arrow_props) ## arrow encompassing data range\n",
    "    \n",
    "    orf_name='%s %s (%d protein sequences)'%(rename[aln] if aln in rename else aln,genes[aln],len(seqs[aln])) ## format ORF name\n",
    "    \n",
    "    ax.text(len(ys[aln])/2,y+arrow_props['width']/2-0.01,orf_name,ha='center',va='center',color='w',size=26) ## add text in the middle\n",
    "    \n",
    "    if aln in ['Anellos','Alphas','Betas','Gammas']:\n",
    "        ax.axvspan(4,150,facecolor='lightgrey',edgecolor='none',alpha=0.6,zorder=0) ## add grey box for anellos where the jelly-roll domain roughly is\n",
    "    \n",
    "    ax.set_ylim(0,24)\n",
    "    ax.set_xlim(0,len(ys[aln])) ## plot bounds\n",
    "\n",
    "    [ax.spines[loc].set_visible(False) for loc in ax.spines if loc not in ['bottom','left']] ## remove right and top spines\n",
    "    \n",
    "    ax.axhline(20.0,ls='--',color='k') ## max amino acid boundary\n",
    "    ax.tick_params(size=5,labelsize=20)\n",
    "    \n",
    "    if column==0:\n",
    "        ax.set_ylabel('unique amino acids',size=24) ## y-axis label for leftmost plots\n",
    "    else:\n",
    "        ax.tick_params(axis='y',size=5,labelsize=0)\n",
    "        \n",
    "    if row==3:\n",
    "        ax.set_xlabel('amino acid site',size=24) ## x-axis label for bottom plots\n",
    "        \n",
    "    ax.grid(axis='y',ls='--',color='grey') ## y-axis grid\n",
    "    \n",
    "plt.savefig(os.path.join(base_path, 'figures', 'figure_S1.pdf'),dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.2 64-bit ('paper_gytis': conda)"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "interpreter": {
   "hash": "6145e1abf69061c5329082bc7ec51bd3f5f08c08c17674746f4f23a8991aef83"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}