{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-23T06:03:45.324607Z",
     "start_time": "2020-12-23T06:03:42.803627Z"
    }
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "\n",
    "import glob,os,re,pathlib\n",
    "\n",
    "typeface='Arial'\n",
    "mpl.rcParams['axes.labelweight']=300\n",
    "mpl.rcParams['font.family']=typeface\n",
    "mpl.rcParams['font.size']=22\n",
    "mpl.rcParams['pdf.fonttype']=42\n",
    "\n",
    "base_path=pathlib.Path(os.path.abspath('')).parent\n",
    "data_path=os.path.join(base_path, \"data\", \"figure_4\")\n",
    "\n",
    "def percent2dissim(matrix):\n",
    "    convert=lambda percent: 1.0-(percent/100.0)\n",
    "    \n",
    "    for i,row in enumerate(matrix):\n",
    "        for j,col in enumerate(matrix):\n",
    "            if i==j:\n",
    "                matrix[row][col]=0.0\n",
    "            elif j>i:\n",
    "                matrix[row][col]=convert(matrix[row][col])\n",
    "                matrix[col][row]=matrix[row][col]\n",
    "    \n",
    "    return matrix\n",
    "\n",
    "distance_matrices={}\n",
    "seq_names={}\n",
    "\n",
    "for aln_path in glob.glob(os.path.join(data_path,'*_aa.fasta')): ## iterate over fastas in folder\n",
    "    aln=SeqIO.parse(aln_path,format='fasta') ## open alignment\n",
    "    vir=os.path.basename(aln_path).split('.')[0].split('_')[0] ## get alignment stem name\n",
    "    \n",
    "    print('%s\\n%s'%(vir,aln_path))\n",
    "    seqs=[]\n",
    "    seq_names[vir]=[]\n",
    "    \n",
    "    for seq in aln:\n",
    "        seq_names[vir].append(seq.id) ## remember names in alignment\n",
    "        seqs.append(seq.seq) ## get sequences in alignment\n",
    "    \n",
    "    matrix_fname=aln_path.replace('.fasta','.p_matrix.tsv') ## generate distance matrix file name\n",
    "    print(matrix_fname)\n",
    "    \n",
    "    if os.path.isfile(matrix_fname): ## matrix file already exists\n",
    "        matrix_file=open(matrix_fname,'r')\n",
    "\n",
    "#             distance_matrices[vir]={}\n",
    "        for line in matrix_file:\n",
    "            l=line.strip('\\n').split(',')\n",
    "            if l[0]=='':\n",
    "                cols=l[1:]\n",
    "                distance_matrices[vir]={a: {b: None for b in cols} for a in cols}\n",
    "                seq_names[vir]=l[1:]\n",
    "            else:\n",
    "                for c in range(len(cols)):\n",
    "                    distance_matrices[vir][l[0]][cols[c]]=float(l[c+1]) if l[c+1]!='' else 100.0\n",
    "\n",
    "        print(len(distance_matrices[vir]))\n",
    "    else: ## matrix file doesn't exist\n",
    "        print('File %s does not exist, computing matrix'%(matrix_fname))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-23T06:03:45.371340Z",
     "start_time": "2020-12-23T06:03:45.328639Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "transmission_pairs={vir: {} for vir in seq_names}\n",
    "\n",
    "for fname in glob.glob(os.path.join(data_path,'*.csv')): ## iterate over CSVs\n",
    "    print(fname)\n",
    "    for vir in seq_names:\n",
    "        for line in open(fname,'r'): ## iterate over lines\n",
    "            l=line.strip('\\n').split(',')\n",
    "            if len(l)>4: ## are we dealing with transmitted or non-transmitted data?\n",
    "                name,coverage,sample,source,d01,d02,d03,d04,d05,d06,t00,t01,t02,t03,t04=l\n",
    "                transmitted='transmitted'\n",
    "            elif len(l)==4:\n",
    "                name,coverage,timepoint,value=l\n",
    "                transmitted='not transmitted'\n",
    "                \n",
    "            if name!='orf': ## not header\n",
    "                name_parts=name.split('|')\n",
    "                contig=name_parts[1].split('_')\n",
    "                name_parts[1]='_'.join(contig[:-3])\n",
    "                modify_name='|'.join(name_parts) ## should now match sequence names\n",
    "                s=name[:3] ## get patient pair id\n",
    "\n",
    "                if s not in transmission_pairs[vir]: transmission_pairs[vir][s]={'transmitted': [], 'not transmitted': [],'resident': []}\n",
    "                \n",
    "                if modify_name in seq_names[vir]:\n",
    "                    transmission_pairs[vir][s][transmitted].append(modify_name)\n",
    "\n",
    "endemic_threshold=0.00 ## no threshold by default\n",
    "study_seq=re.compile('^R[0-9]+[DT][0-9]+')\n",
    "t0=re.compile('^R[0-9]+T00')\n",
    "\n",
    "\n",
    "for vir in transmission_pairs: ## iterate over datasets\n",
    "    for sample in transmission_pairs[vir]: ## iterate over donor-recipient pair\n",
    "        endemics=[]\n",
    "        for r_idx,recipient_anello in enumerate(seq_names[vir]): ## iterate over all names in dataset\n",
    "            if recipient_anello.startswith(sample): ## name belongs to donor-recipient pair\n",
    "#                 print(recipient_anello)\n",
    "                if t0.match(recipient_anello): ## virus in recipient\n",
    "                    print(vir,recipient_anello)\n",
    "                    close=False\n",
    "                    for transmitted_anello in transmission_pairs[vir][sample]['transmitted']: ## iterate over transmitted anellos in pair\n",
    "                        \n",
    "                        dist=distance_matrices[vir][recipient_anello][transmitted_anello]\n",
    "                        if dist<endemic_threshold: ## compare distance of anello in recipient to transmitted anello - if it's above a similarity threshold it's a likely endemic\n",
    "                            close=True\n",
    "                    \n",
    "                    if close==False:\n",
    "                        transmission_pairs[vir][sample]['resident'].append(recipient_anello)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-23T06:03:52.276633Z",
     "start_time": "2020-12-23T06:03:45.374614Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "colours={'transmitted': 'indianred','not transmitted': 'slategrey','resident': 'seagreen'}\n",
    "categories=['transmitted','not transmitted','resident']\n",
    "\n",
    "comparisons={a: {b: [] for b in categories} for a in categories}\n",
    "\n",
    "for vir in transmission_pairs:\n",
    "    \n",
    "    if vir in ['Alphas','Betas','Gammas']:\n",
    "        distr={}\n",
    "        count=0\n",
    "\n",
    "        labels=[]\n",
    "        store_label=''\n",
    "        for pair in [(a,b) for a in categories for b in categories if categories.index(a)<=categories.index(b)]:\n",
    "            type1,type2=pair\n",
    "            for sample in transmission_pairs[vir]:\n",
    "                for A in transmission_pairs[vir][sample][type1]:\n",
    "                    for B in transmission_pairs[vir][sample][type2]:\n",
    "                        if A!=B:\n",
    "                            comparisons[type1][type2].append(distance_matrices[vir][A][B])\n",
    "\n",
    "fig = plt.figure(figsize=(15, 10),facecolor='w') \n",
    "gs = gridspec.GridSpec(1,1,wspace=0.01,hspace=0.1) \n",
    "ax=plt.subplot(gs[0],facecolor='w')\n",
    "\n",
    "count=0\n",
    "store_label=''\n",
    "for pair in [(a,b) for a in categories for b in categories if categories.index(a)<=categories.index(b)]:\n",
    "    type1,type2=pair\n",
    "    \n",
    "    kde_fc=colours[type2]\n",
    "    label_fc=colours[type1]\n",
    "    labels.append('%s'%(type2))\n",
    "    kde=gaussian_kde(comparisons[type1][type2])\n",
    "    x_grid=np.linspace(min(comparisons[type1][type2]),max(comparisons[type1][type2]),100)\n",
    "    ys_raw=kde(x_grid)\n",
    "    ys=ys_raw/ys_raw.max()/2.1\n",
    "    \n",
    "    ax.fill_betweenx(x_grid,[count-y for y in ys],[count+y for y in ys],facecolor=kde_fc,alpha=0.1,zorder=9)\n",
    "    ax.plot([count-y for y in ys],x_grid,color='k',zorder=10)\n",
    "    ax.plot([count+y for y in ys],x_grid,color='k',zorder=10)\n",
    "\n",
    "    if store_label!=type1:\n",
    "        ax.axvline(count-0.5,color='k',lw=2,ls='--')\n",
    "        ax.text(count-0.4,100.0,'%s versus'%(type1),size=24,ha='left',va='bottom',color=label_fc)\n",
    "    \n",
    "    xs=[count+np.random.uniform(-kde(c)/ys_raw.max()/2.1,kde(c)/ys_raw.max()/2.1) for c in comparisons[type1][type2]]\n",
    "\n",
    "    ax.scatter(xs,comparisons[type1][type2],s=5,facecolor=kde_fc,alpha=1.0,zorder=1)\n",
    "    \n",
    "    mean=np.mean(comparisons[type1][type2])\n",
    "    median=np.median(comparisons[type1][type2])\n",
    "    print('%s\\t%s\\t%.4f\\t%.3f\\t%s'%(type1,type2,median,np.std(comparisons[type1][type2]),len(comparisons[type1][type2])))\n",
    "    ax.plot([count-0.4,count+0.4],[mean,mean],color='k',zorder=1000)\n",
    "    ax.plot([count-0.4,count+0.4],[median,median],color='k',ls='--',zorder=1000)\n",
    "\n",
    "    count+=1\n",
    "    store_label=type1\n",
    "\n",
    "[ax.spines[loc].set_visible(False) for loc in ax.spines if loc not in ['left']]\n",
    "ax.set_ylabel('percent amino acid identity',size=30)\n",
    "ax.tick_params(axis='x',size=0,labelsize=24)\n",
    "ax.tick_params(axis='y',labelsize=24)\n",
    "\n",
    "ax.set_xticks(range(count))\n",
    "ax.set_xticklabels(labels,size=16)\n",
    "\n",
    "ax.grid(axis='y',ls='--',zorder=0)\n",
    "ax.set_ylim(-5,105)\n",
    "plt.savefig(os.path.join(base_path, 'figures', 'figure_4.pdf'),dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.2 64-bit ('paper_gytis': conda)"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "interpreter": {
   "hash": "6145e1abf69061c5329082bc7ec51bd3f5f08c08c17674746f4f23a8991aef83"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}