{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-23T15:35:49.946795Z",
     "start_time": "2020-11-23T15:35:48.788624Z"
    }
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "\n",
    "from Bio import SeqIO,Phylo\n",
    "import numpy as np\n",
    "\n",
    "import glob,os,re,copy,math,pathlib\n",
    "\n",
    "typeface='Arial'\n",
    "mpl.rcParams['axes.labelweight']=300\n",
    "mpl.rcParams['font.family']=typeface\n",
    "mpl.rcParams['font.size']=22\n",
    "mpl.rcParams['pdf.fonttype']=42\n",
    "\n",
    "import importlib.machinery\n",
    "import importlib.util\n",
    "import baltic as bt\n",
    "\n",
    "loader = importlib.machinery.SourceFileLoader('baltic', os.path.join(os.getcwd(), 'baltic_utils.py'))\n",
    "spec = importlib.util.spec_from_loader(loader.name, loader)\n",
    "bt_utils = importlib.util.module_from_spec(spec)\n",
    "loader.exec_module(bt_utils)\n",
    "\n",
    "base_path=pathlib.Path(os.path.abspath('')).parent\n",
    "data_path=os.path.join(base_path, \"data\", \"figure_5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-23T15:35:50.842425Z",
     "start_time": "2020-11-23T15:35:49.952295Z"
    }
   },
   "outputs": [],
   "source": [
    "cfml_path='ClonalFrameML'\n",
    "\n",
    "homoplasy_trees={}\n",
    "\n",
    "for tree_path in glob.glob(os.path.join(data_path,'anello_80.0_cluster*_*.newick')):\n",
    "    if 'chunk' not in tree_path and 'rename' not in tree_path and 'CFML' not in tree_path:\n",
    "        \n",
    "        print(tree_path)\n",
    "        aln_path=tree_path.replace('.newick','.fasta')\n",
    "        out_stem=tree_path.replace('.newick','_CFML')\n",
    "        bt_utils.reconstruct_ancestral_states(cfml_path,aln_path,tree_path,2.0,out_stem)\n",
    "        \n",
    "        genus=tree_path.split('_')[-1].split('.')[0]\n",
    "        \n",
    "        ll=bt.loadNexus(tree_path.replace('.newick','_CFML.annotated.tree'),verbose=False)\n",
    "        ll.treeStats()\n",
    "        \n",
    "        for k in ll.Objects:\n",
    "            if 'sequence' in k.parent.traits:\n",
    "                pseq=k.parent.traits['sequence']\n",
    "                seq=k.traits['sequence']\n",
    "                for site,pair in enumerate(zip(pseq,seq)):\n",
    "                    a,b=pair\n",
    "                    invalid=['N','-']\n",
    "                    if a!=b and a not in invalid and b not in invalid:\n",
    "                        if 'mutations' not in k.traits: k.traits['mutations']=[]\n",
    "                        k.traits['mutations'].append('%s%d%s'%(a,site,b))\n",
    "        \n",
    "        homoplasy_trees[genus]=ll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-23T15:35:51.228766Z",
     "start_time": "2020-11-23T15:35:50.847483Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "homoplasy_branches={}\n",
    "synapomorphy_branches={}\n",
    "\n",
    "for genus in homoplasy_trees:\n",
    "    ll=homoplasy_trees[genus]\n",
    "    homoplasy_branches[genus]={}\n",
    "    synapomorphy_branches[genus]={}\n",
    "    \n",
    "    all_mutations=Counter(sum(ll.getParameter('mutations',use_trait=True),[])) ## counts of all mutations across the tree\n",
    "    \n",
    "    print(genus)\n",
    "    for mu in all_mutations: ## for each mutation\n",
    "        mutation_branches=ll.getBranches(lambda w: 'mutations' in w.traits and mu in w.traits['mutations']) ## get all branches that have this mutation\n",
    "        if isinstance(mutation_branches,list)==False: mutation_branches=[mutation_branches]\n",
    "        if all_mutations[mu]>1: ## if it occurs more than once\n",
    "            for branch in mutation_branches: ## iterate over each member\n",
    "                if mu not in homoplasy_branches[genus]: homoplasy_branches[genus][mu]=[]\n",
    "                homoplasy_branches[genus][mu].append(branch)\n",
    "        else:\n",
    "            synapomorphy_branches[genus][mu]=mutation_branches[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-23T15:36:07.792479Z",
     "start_time": "2020-11-23T15:35:51.262912Z"
    }
   },
   "outputs": [],
   "source": [
    "def split_aln(sequence_list,out_name,start,end):\n",
    "    \"\"\"\n",
    "    Given an alignment and coordinates output a chunk of the alignment.\n",
    "    \"\"\"\n",
    "    out=open(out_name,'w')\n",
    "    chunk=[]\n",
    "    for seq in sequence_list:\n",
    "        seq_copy=copy.deepcopy(seq)\n",
    "        seq_copy.seq=seq.seq[start:end]\n",
    "        chunk.append(seq_copy)\n",
    "\n",
    "    SeqIO.write(chunk,out,format='phylip')\n",
    "    out.close()\n",
    "\n",
    "def midpoint_root(tree_path,out_name,rename=None):\n",
    "    \"\"\"\n",
    "    Load tree, midpoint root, output back.\n",
    "    \"\"\"\n",
    "    newick=Phylo.read(open(tree_path,'r'),format='newick')\n",
    "    newick.root_at_midpoint()\n",
    "    \n",
    "    if rename:\n",
    "        for tip in newick.get_terminals():\n",
    "            tip.name=rename[int(tip.name)]\n",
    "\n",
    "    out=open(out_name,'w')\n",
    "    Phylo.write(newick,out,format='newick')\n",
    "    out.close()\n",
    "\n",
    "for aln in glob.glob(os.path.join(data_path,'anello*.fasta')):\n",
    "    print(aln)\n",
    "    if 'chunk' not in aln and 'ML_sequence' not in aln and 'rename' not in aln and 'reconstructed' not in aln:\n",
    "        clust_id=re.compile('anello\\_([0-9\\.]+)\\_(cluster[0-9]+)\\_[A-Za-z]+\\.fasta')\n",
    "        cluster_id=clust_id.search(aln).group(2)\n",
    "        \n",
    "        seqs=[]\n",
    "        seq_name={}\n",
    "        for s,seq in enumerate(SeqIO.parse(aln,format='fasta')):\n",
    "            seq_name[s]=seq.id\n",
    "            seq.id=str(s)\n",
    "            seqs.append(seq)\n",
    "            \n",
    "            L=len(seq.seq)\n",
    "\n",
    "        print(L)\n",
    "        chunks=range(0,L,500)\n",
    "        for i,j in zip(chunks,chunks[1:]): ## iterate through chunks\n",
    "            out_name=aln.replace('.fasta','_chunk_%04d-%04d.phylip'%(i,j))\n",
    "            split_aln(seqs,out_name,i,j) ## save alignment chunk\n",
    "            os.system('phyml -i %s -b 0 -m HKY85 -f m -t e -v 0 -c 4 -a e -o tlr'%(out_name)) ## build phyml tree\n",
    "            \n",
    "            tree_name='%s_phyml_tree.txt'%(out_name)\n",
    "            out_tree=tree_name.replace('.phylip_phyml_tree.txt','.newick')\n",
    "            print(tree_name)\n",
    "            print(out_name)\n",
    "            \n",
    "            midpoint_root(tree_name,out_tree,seq_name) ## midpoint root tree, rename tips to original names, save to file\n",
    "            os.system('rm %s'%(tree_name)) ## remove phyml tree\n",
    "            os.system('rm %s'%(tree_name.replace('_phyml_tree','_phyml_stats')))\n",
    "            \n",
    "        out_name=aln.replace('.fasta','_chunk_%04d-%04d.phylip'%(j,L))\n",
    "        split_aln(seqs,out_name,j,L)\n",
    "        os.system('phyml -i %s -b 0 -m HKY85 -f m -t e -v 0 -c 4 -a e -o tlr'%(out_name))\n",
    "\n",
    "        tree_name='%s_phyml_tree.txt'%(out_name)\n",
    "        print(tree_name)\n",
    "        print(out_name)\n",
    "        out_tree=tree_name.replace('.phylip_phyml_tree.txt','.newick')\n",
    "\n",
    "     \n",
    "        midpoint_root(tree_name,out_tree,seq_name)\n",
    "        os.system('rm %s'%(tree_name))\n",
    "        os.system('rm %s'%(tree_name.replace('_phyml_tree','_phyml_stats')))\n",
    "\n",
    "print('Done!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-23T15:36:07.863530Z",
     "start_time": "2020-11-23T15:36:07.794938Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "genus=lambda file_name: re.compile('cluster[0-9]+_([A-Za-z]+)_chunk').search(file_name).group(1)\n",
    "\n",
    "tanglegram_trees={}\n",
    "\n",
    "for tree_path in sorted(glob.glob(os.path.join(data_path,'anello_*_chunk*.newick'))):\n",
    "    print(tree_path)\n",
    "    g=genus(tree_path)\n",
    "    if g not in tanglegram_trees: tanglegram_trees[g]=[]\n",
    "    \n",
    "    ll=bt.loadNewick(tree_path)\n",
    "    ll.sortBranches()\n",
    "    ll=ll.collapseBranches(lambda w: w.length==0.0)\n",
    "    tanglegram_trees[g].append(ll)\n",
    "    print(len(ll.getExternal()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-23T15:40:58.539179Z",
     "start_time": "2020-11-23T15:36:07.867445Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\n",
    "def valid_haplotypes(A,B):\n",
    "    \"\"\"\n",
    "    Given two polymorphic sites remove gap and ambiguity characters, return if both sites still polymorphic and at least 10% of original sample size\n",
    "    \"\"\"\n",
    "    assert len(A)==len(B) ## can create haplotypes out of columns\n",
    "    \n",
    "    filtered_haps=[h for h in zip(A,B) if '-' not in h and 'N' not in h] ## filter haplotypes that contain N or gap\n",
    "    two_columns=tuple(zip(*filtered_haps)) ## unpack columns\n",
    "    \n",
    "    if len(filtered_haps)>len(A)*0.1 and min([len(set(allele)) for allele in two_columns])>1: ## filtered haplotypes are at least 10% of all haplotypes and both sites are still polymorphic\n",
    "        return two_columns\n",
    "\n",
    "def haplotype_freqs(A,B):\n",
    "    \"\"\"\n",
    "    Given two polymorphic sites compute the frequencies of alleles and haplotypes\n",
    "    \"\"\"\n",
    "    As=Counter(A) ## count alleles at site 1\n",
    "    Bs=Counter(B) ## same for site 2\n",
    "    haps=Counter(tuple(zip(A,B))) ## count haplotypes between the two\n",
    "    total=sum(haps.values()) ## total number of samples\n",
    "    \n",
    "    aF={a: As[a]/total for a in As} ## frequency of each allele\n",
    "    bF={b: Bs[b]/total for b in Bs}\n",
    "    \n",
    "    hapC={(a,b): haps[(a,b)] if (a,b) in haps else 0.0 for a in As for b in Bs} ## frequency of each possible haplotype\n",
    "    \n",
    "    return aF,bF,hapC\n",
    "\n",
    "def chiSqDf(aF,bF,hapF):\n",
    "    \"\"\"\n",
    "    Given observed allele and haplotype frequencies compute LD between two sites\n",
    "    \"\"\"\n",
    "    total=sum(hapF.values())\n",
    "    O,E=zip(*((hapF[(a,b)],aF[a]*bF[b]*total) for a in aF for b in bF)) ## get observed and expected haplotype counts\n",
    "    chisq=0.0\n",
    "    \n",
    "    for o,e in zip(O,E): ## iterate over each haplotype's contribution \n",
    "        chisq+=math.pow(o-e,2)/e ## incremend with ((observed-expected)^2)/expected\n",
    "    \n",
    "    df=((len(aF)-1)*(len(bF)-1))\n",
    "    chisqdf=chisq/df/total ## normalise by degrees of freedom and total number of samples\n",
    "    return chisqdf\n",
    "\n",
    "filter2valid=lambda k: list(filter(lambda w: w in ['A','C','T','G'],k)) ## only keep values that are valid nucleotides\n",
    "\n",
    "keep={}\n",
    "aln={}\n",
    "\n",
    "for aln_path in glob.glob(os.path.join(data_path,'*_nt.fasta')): ## iterate over alignments\n",
    "    if 'chunk' not in aln_path and 'renamed' not in aln_path and 'CFML' not in aln_path:\n",
    "        genus=os.path.basename(aln_path).split('_')[-2]\n",
    "        if genus in ['Alpha','Beta','Gamma']:\n",
    "            aln[genus]={}\n",
    "            keep[genus]={}\n",
    "            for seq in SeqIO.parse(aln_path,format='fasta'): ## go over sequences in alignment\n",
    "                aln[genus][seq.id]=list(seq.seq)\n",
    "                L=len(seq.seq)\n",
    "\n",
    "            keep[genus]=[] ## remember which sites to keep\n",
    "            for i in range(L):\n",
    "                raw_column=[site[i] for site in [aln[genus][s] for s in sorted(aln[genus])]] ## get column values\n",
    "                filt_column=filter2valid(raw_column) ## filter column down to valid residues\n",
    "\n",
    "                if len(raw_column)*0.1<len(filt_column): ## if at least 10% of column is good - keep\n",
    "                    keep[genus].append(i)\n",
    "\n",
    "            for seq in aln[genus]:\n",
    "                aln[genus][seq]=''.join([aln[genus][seq][i] for i in keep[genus]]) ## trim sequences\n",
    "\n",
    "columns={}\n",
    "\n",
    "filter2valid=lambda k: list(filter(lambda w: w in ['A','C','T','G'],k))\n",
    "\n",
    "for genus in keep: ## iterate over alignments\n",
    "    columns[genus]={}\n",
    "    \n",
    "    for site in range(len(keep[genus])): ## iterate over sites to keep\n",
    "        raw_column=[seq[site] for seq in [aln[genus][s] for s in sorted(aln[genus])]]\n",
    "        filt_column=filter2valid(raw_column) ## filter column to valid residues\n",
    "        col_freq=Counter(filt_column) ## get allele frequencies in column\n",
    "        counts=col_freq.values() ## get counts of each allele\n",
    "        min_freq,tot_freq=min(counts),sum(counts) ## get minor allele count and total number of valid sequences\n",
    "\n",
    "        more_than_10_percent = len(filt_column) > len(raw_column)*0.1 ## boolean of whether at least 10% sequences are valid at site\n",
    "        minor_allele = min_freq/tot_freq > 0.05 ## boolean of whether minor allele frequency is at least 5%\n",
    "        multi_allelic = len(col_freq)>1 ## boolean that site contains at least 2 alleles\n",
    "        \n",
    "        if more_than_10_percent and minor_allele and multi_allelic:\n",
    "            columns[genus][site]=raw_column ## remember raw column\n",
    "\n",
    "LD={}\n",
    "\n",
    "for genus in columns:\n",
    "    LD[genus]={siteA: {siteB: np.nan for siteB in columns[genus] if siteB>siteA} for siteA in columns[genus]} ## matrix for linkage disequilibrium values\n",
    "    print(genus)\n",
    "    for siteA in columns[genus]: ## iterate over sites\n",
    "        for siteB in columns[genus]: ## iterate over sites again\n",
    "            if siteB>siteA: ## only care about upper triangle\n",
    "                good_haplotype=valid_haplotypes(columns[genus][siteA],columns[genus][siteB]) ## confirm haplotypes produced by the two columns are valid\n",
    "                if good_haplotype: ## if all good\n",
    "                    c=round(chiSqDf(*haplotype_freqs(*good_haplotype)),4) ## compute LD\n",
    "                    LD[genus][siteA][siteB]=c ## store LD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-11-23T15:43:17.175238Z",
     "start_time": "2020-11-23T15:40:58.562046Z"
    }
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(30, 20),facecolor='w') ## start plot\n",
    "gs = gridspec.GridSpec(3,3,height_ratios=[1,1,1],width_ratios=[2,1,1],wspace=0.1,hspace=0.12) \n",
    "\n",
    "cmap=mpl.cm.twilight\n",
    "\n",
    "chunk_colours={(0,500): '#7fcdbb', \n",
    "               (500,1000): '#41b6c4', \n",
    "               (1000,1500): '#1d91c0', \n",
    "               (1500,2000): '#225ea8', \n",
    "               (2000,2300): '#0c2c84'}\n",
    "\n",
    "for g,G in enumerate(sorted(tanglegram_trees)): ## iterate over genera\n",
    "    print(G)\n",
    "    \n",
    "    tanglegram_cmap=mpl.cm.RdYlGn\n",
    "    branch_cmap=mpl.cm.cividis\n",
    "    \n",
    "    untangled=tanglegram_trees[G]\n",
    "    \n",
    "    ax=plt.subplot(gs[g,0],facecolor='w')\n",
    "    \n",
    "    offset=0\n",
    "    gap=0.2\n",
    "    tip_positions={}\n",
    "    for t,tr in enumerate(untangled): ## iterate over trees\n",
    "        tr.sortBranches()\n",
    "        x_attr=lambda k: k.height + offset ## offset tree\n",
    "\n",
    "        tip_positions[tr]={k.name: (x_attr(k),k.y) for k in tr.getExternal()}\n",
    "        \n",
    "        fc=lambda k: tanglegram_cmap(tip_positions[untangled[0]][k.name][1]/tr.ySpan)\n",
    "        s=20+(1000/len(tr.getExternal()))\n",
    "        \n",
    "        tr.plotTree(ax,x_attr=x_attr,width=5,colour=[chunk_colours[c] for c in sorted(chunk_colours,key=lambda w: w[0])][t]) ## plot tree\n",
    "        tr.plotPoints(ax,x_attr=x_attr,size=s,colour=fc)\n",
    "        \n",
    "        sorted_chunks=sorted(chunk_colours,key=lambda w: w[0])\n",
    "        x=sorted_chunks[t][0]/2500\n",
    "        y=0.95\n",
    "        dx=(sorted_chunks[t][1]-sorted_chunks[t][0])/2500\n",
    "        dy=0\n",
    "        w=0.08\n",
    "        hL=0.0\n",
    "        if t==len(untangled)-1: ## if at last chunk make arrow pointed\n",
    "            hL=40/2500\n",
    "        \n",
    "        ax.arrow(x,y,dx,dy,facecolor=chunk_colours[sorted_chunks[t]],edgecolor='none', \n",
    "                 width=w,length_includes_head=True,head_length=hL,head_width=w,transform=ax.transAxes)\n",
    "        \n",
    "        offset+=tr.treeHeight+gap\n",
    "    \n",
    "    for this_tree,next_tree in zip(untangled,untangled[1:]): ## iterate over pairs of adjacent trees\n",
    "        for k in this_tree.getExternal():\n",
    "            x1,y1=tip_positions[this_tree][k.name]\n",
    "            x2,y2=tip_positions[next_tree][k.name]\n",
    "\n",
    "            matching_tip=next_tree.getBranches(lambda w: w.branchType=='leaf' and w.name==k.name)\n",
    "\n",
    "            fr=0.3\n",
    "            fc=tanglegram_cmap(tip_positions[untangled[0]][matching_tip.name][1]/this_tree.ySpan)\n",
    "            highest_tip=max(tip_positions[this_tree].values())\n",
    "            \n",
    "            lw=3+(50/len(tip_positions[this_tree].keys()))\n",
    "            \n",
    "            ax.plot([x1,highest_tip[0]+fr*gap,x2-matching_tip.height-fr*gap,x2],[y1,y1,y2,y2],color=fc,lw=lw,alpha=1.0,zorder=0)\n",
    "    \n",
    "    ax.plot()\n",
    "    ax.tick_params(labelsize=0,size=0)\n",
    "    [ax.spines[loc].set_visible(False) for loc in ax.spines]\n",
    "    \n",
    "    ax.set_xlim(-0.06,offset-gap+0.05)\n",
    "    ax.set_ylim(-tr.ySpan*0.01,tr.ySpan*1.15)\n",
    "    \n",
    "    if G=='Alpha':\n",
    "        ax.text(0.0,1.1,'A',size=50,ha='left',va='bottom',transform=ax.transAxes)\n",
    "        \n",
    "    ax.text(0.01,1.0,'%storquevirus'%(G),size=30,ha='left',va='bottom',transform=ax.transAxes)\n",
    "    \n",
    "    ########################\n",
    "    ax=plt.subplot(gs[g,1],facecolor='w')\n",
    "    \n",
    "    ll=homoplasy_trees[G]\n",
    "    \n",
    "    first_tree_ys={k.name: k.y for k in tanglegram_trees[G][0].getExternal()}\n",
    "    \n",
    "    fc=lambda k: tanglegram_cmap(first_tree_ys[k.name]/ll.ySpan) if k.name in first_tree_ys else 'grey'\n",
    "    \n",
    "    ll.plotTree(ax,colour='k',width=5) ## plot tree\n",
    "    \n",
    "    s=20+(3000/len(ll.getExternal())) ## tip size depends on total number of tips in tree\n",
    "    ll.plotPoints(ax,colour=fc,size=s) ## colour tips based on their y position in the first tree\n",
    "    \n",
    "    homoplasy_lines=[]\n",
    "    \n",
    "    xs={t: [] for t in chunk_colours}\n",
    "    ys={t: [] for t in chunk_colours}\n",
    "    for mu in synapomorphy_branches[G]:\n",
    "        k=synapomorphy_branches[G][mu]\n",
    "        site=int(mu[1:-1])\n",
    "        fromM=mu[0]\n",
    "        toM=mu[-1]\n",
    "        \n",
    "        frac=site/len(k.traits['sequence'])\n",
    "        for interval in chunk_colours:\n",
    "            i,j=interval\n",
    "            if i<site<=j:\n",
    "                xs[interval].append(k.parent.x+k.length*frac)\n",
    "                ys[interval].append(k.y)\n",
    "    \n",
    "    for interval in chunk_colours:\n",
    "        fc=chunk_colours[interval]\n",
    "        ax.scatter(xs[interval],ys[interval],s=80,facecolor=fc,lw=3,marker='|',zorder=100)\n",
    "        ax.scatter(xs[interval],ys[interval],s=90,facecolor='k',lw=5,marker='|',zorder=99)\n",
    "    \n",
    "    max_homoplasy=max(map(len,homoplasy_branches[G].values()))\n",
    "    for mu in homoplasy_branches[G]: ## iterate over mutations on homoplasious branches\n",
    "        sorted_branches=sorted(homoplasy_branches[G][mu],key=lambda w: w.height) ## all branches with same mutation\n",
    "        for b1,b2 in zip(sorted_branches,sorted_branches[1:]): ## only do pairs of branches, not even all branches (because space is limited)\n",
    "            site=int(mu[1:-1])\n",
    "            fromM=mu[0]\n",
    "            toM=mu[-1]\n",
    "\n",
    "            frac=site/len(b1.traits['sequence'])\n",
    "            x1=b1.parent.x+b1.length*frac\n",
    "            y1=b1.y\n",
    "            \n",
    "            x2=b2.parent.x+b2.length*frac\n",
    "            y2=b2.y\n",
    "            \n",
    "            for interval in chunk_colours:\n",
    "                i,j=interval\n",
    "                if i<site<=j: ## get interval colour\n",
    "                    fc=chunk_colours[interval]\n",
    "            ax.plot([x1,x2],[y1,y2],color=fc,alpha=1.0,zorder=0)\n",
    "    \n",
    "    ax.tick_params(axis='y',labelsize=0,size=0)\n",
    "    ax.tick_params(axis='x',labelsize=28)\n",
    "    [ax.spines[loc].set_visible(False) for loc in ax.spines if loc not in ['bottom']]\n",
    "    \n",
    "    if G=='Gamma':\n",
    "        ax.set_xlabel('divergence (nt subs/site)',size=26)\n",
    "    \n",
    "    elif G=='Alpha':\n",
    "        ax.text(0.0,1.1,'B',size=50,ha='left',va='bottom',transform=ax.transAxes)\n",
    "    \n",
    "    ax.set_xlim(-0.005,ll.treeHeight+0.01)\n",
    "    ax.set_ylim(-ll.ySpan*0.01,ll.ySpan*1.05)\n",
    "    \n",
    "    #########################\n",
    "    ax=plt.subplot(gs[g,2],facecolor='w')\n",
    "    pairs=[(i,j) for i in columns[G] for j in columns[G] if j>i and np.isnan(LD[G][i][j])==False] ## get all pairs of sites that have measured LD\n",
    "\n",
    "    xs=np.array(list(map(np.diff,pairs)))\n",
    "    xs=xs.reshape(len(xs))\n",
    "\n",
    "    ys=np.array([LD[G][x[0]][x[1]] for x in pairs])\n",
    "    \n",
    "    bins=np.arange(0,max(xs),100) ## bin LD values according to distance\n",
    "    binned={i: [] for i in range(len(bins)+1)}\n",
    "    idx=np.digitize(xs,bins)\n",
    "    for i,x in enumerate(idx):\n",
    "        binned[x].append(ys[i])\n",
    "        \n",
    "    ax.step(bins,[np.mean(binned[i]) for i in range(len(bins))],color='indianred',lw=5,zorder=100) ## plot mean of each bin\n",
    "    ax.scatter(xs,ys,s=20,c='k',edgecolor='none',alpha=1.0) ## plot all LD data\n",
    "    \n",
    "    ax.yaxis.set_major_locator(mpl.ticker.MultipleLocator(0.2))\n",
    "    \n",
    "    ax.set_xlim(-10,max(xs)+10)\n",
    "    ax.set_ylim(-0.01,1.01)\n",
    "    \n",
    "    ax.grid(axis='y',ls='--',color='grey')\n",
    "    [ax.spines[loc].set_visible(False) for loc in ax.spines if loc in ['top','left']]\n",
    "    ax.tick_params(labelsize=30)\n",
    "    \n",
    "    ax.yaxis.tick_right()\n",
    "    ax.yaxis.set_label_position('right')\n",
    "    \n",
    "    if G=='Gamma':\n",
    "        ax.set_ylabel(r'LD ($\\chi^{2}_{df}$)',size=30)\n",
    "        ax.set_xlabel('distance between sites',size=30)\n",
    "    else:\n",
    "        ax.tick_params(axis='y',size=0,labelsize=0)\n",
    "        if G=='Alpha':\n",
    "            ax.text(0.0,1.1,'C',size=50,ha='left',va='bottom',transform=ax.transAxes)\n",
    "            \n",
    "plt.savefig(os.path.join(base_path, 'figures', 'figure_4_recombination.pdf'),dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}