{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T09:05:15.647982Z",
     "start_time": "2020-10-20T09:05:14.987409Z"
    }
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import pathlib \n",
    "\n",
    "import matplotlib as mpl\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "typeface='Arial'\n",
    "mpl.rcParams['font.family']=typeface\n",
    "mpl.rcParams['font.size']=22\n",
    "mpl.rcParams['pdf.fonttype']=42\n",
    "\n",
    "import importlib.machinery\n",
    "import importlib.util\n",
    "import baltic as bt\n",
    "\n",
    "loader = importlib.machinery.SourceFileLoader('baltic', os.path.join(os.getcwd(), 'baltic_utils.py'))\n",
    "spec = importlib.util.spec_from_loader(loader.name, loader)\n",
    "bt_utils = importlib.util.module_from_spec(spec)\n",
    "loader.exec_module(bt_utils)\n",
    "\n",
    "base_path=pathlib.Path(os.path.abspath('')).parent\n",
    "data_path=pathlib.Path(base_path).joinpath(\"data\", \"figure_S5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T09:05:15.826567Z",
     "start_time": "2020-10-20T09:05:15.650392Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cfml_path='ClonalFrameML'\n",
    "\n",
    "stem='alpha_noncoding'\n",
    "aln_path=str(pathlib.Path(data_path).joinpath('%s.fasta'%(stem)))\n",
    "tree_path=str(pathlib.Path(data_path).joinpath('%s.newick'%(stem)))\n",
    "kappa=2.0\n",
    "out_stem=stem\n",
    "bt_utils.reconstruct_ancestral_states(cfml_path,str(aln_path),str(tree_path),kappa,out_stem)\n",
    "ll=bt.loadNewick(tree_path)\n",
    "ll.treeStats()\n",
    "\n",
    "bt_utils.annotate_tree(ll,tree_path.replace('.newick','.labelled_tree.newick'),aln_path.replace('.fasta','.reconstructed.fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T09:05:18.199398Z",
     "start_time": "2020-10-20T09:05:15.829444Z"
    }
   },
   "outputs": [],
   "source": [
    "all_mutations=set(sum([[mu for mu in k.traits['muts']] for k in ll.Objects if 'muts' in k.traits],[]))\n",
    "site_mutations={int(mu[1:-1]): [M for M in all_mutations if mu[1:-1]==M[1:-1]] for mu in all_mutations}\n",
    "\n",
    "mu_branches={}\n",
    "for site in site_mutations:\n",
    "    mu_branches[site]={}\n",
    "    for mut in site_mutations[site]:\n",
    "        B=ll.getBranches(lambda k: 'muts' in k.traits and mut in k.traits['muts'] and mut[-1]!='-' and mut[0]!='N' and mut[0]!='-' and mut[-1]!='N',warn=False)\n",
    "        if isinstance(B,list)==False:\n",
    "            B=[B]\n",
    "        mu_branches[site][mut]=B\n",
    "\n",
    "homoplasy_degrees={site: {mut: len(mu_branches[site][mut]) for mut in mu_branches[site]} for site in mu_branches}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T09:06:04.103810Z",
     "start_time": "2020-10-20T09:05:18.219085Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def break_up(sites,threshold):\n",
    "    segments=[]\n",
    "    sorted_sites=sorted(sites,key=lambda w: int(w[1:-1]))\n",
    "    \n",
    "    local=set()\n",
    "    \n",
    "    for i in range(1,len(sorted_sites)):\n",
    "        cur_site=int(sorted_sites[i][1:-1])\n",
    "        pre_site=int(sorted_sites[i-1][1:-1])\n",
    "        \n",
    "        within_threshold=(cur_site-pre_site)<=threshold\n",
    "        \n",
    "        if within_threshold: ## next site within range of previous\n",
    "            local.add(sorted_sites[i])\n",
    "            local.add(sorted_sites[i-1])\n",
    "        else: ## outside threshold but not at last value\n",
    "            local.add(sorted_sites[i-1])\n",
    "            segments.append(local)\n",
    "            local=set()\n",
    "            \n",
    "    if len(local)>0:\n",
    "        segments.append(local)\n",
    "        \n",
    "    return segments\n",
    "\n",
    "\n",
    "fig = plt.figure(figsize=(50, 15),facecolor='w') \n",
    "gs = gridspec.GridSpec(1,2,width_ratios=[1,10],wspace=0.01) \n",
    "\n",
    "ax=plt.subplot(gs[0],facecolor='w')\n",
    "ax2=plt.subplot(gs[1],facecolor='w')\n",
    "\n",
    "ll.traverse_tree()\n",
    "\n",
    "ll.plotTree(ax)\n",
    "ax.set_xlabel('divergence (nt)\\nsubs/site',size=30)\n",
    "ax.tick_params(axis='x',labelsize=26)\n",
    "ax.plot()\n",
    "\n",
    "from collections import Counter\n",
    "from matplotlib.patches import Rectangle\n",
    "mut_freq=Counter(sum([k.traits['muts'] for k in ll.Objects if 'muts' in k.traits],[]))\n",
    "\n",
    "i=0\n",
    "\n",
    "xs=[]\n",
    "ys=[]\n",
    "ms=[]\n",
    "import time\n",
    "\n",
    "y_coords={k.name: k.y for k in ll.getExternal()}\n",
    "colour={'A': 'indianred', 'C': 'steelblue', 'G': 'yellow', 'T': 'seagreen', '-': 'grey', 'N': 'grey'}\n",
    "\n",
    "mutated_branches=[k for k in ll.Objects if 'muts' in k.traits]\n",
    "positions=set()\n",
    "plot_mus={}\n",
    "\n",
    "mu_good=lambda mu: '-' not in mu and 'N' not in mu\n",
    "tract_cutoff=3\n",
    "\n",
    "for k in mutated_branches: ## iterate over branches\n",
    "    muts1=set([m for m in k.traits['muts'] if mut_freq[m]>1 and mu_good(m)]) ## only homoplasies\n",
    "    \n",
    "    for w in mutated_branches: ## iterate over other branches\n",
    "        if w!=k: ## not dealing with same branch and branch has mutations\n",
    "            muts2=set([m for m in w.traits['muts'] if mut_freq[m]>1 and mu_good(m)]) ## get homoplasies on other branch\n",
    "                \n",
    "            if len(muts2.intersection(muts1))>=tract_cutoff: ## at least two mutations shared between branches\n",
    "                for tract in break_up(muts2.intersection(muts1),10): ## break up mutations into chunks within 10 nt of each other\n",
    "                    if len(tract)>=tract_cutoff: ## tract with at least 4 mutations\n",
    "                        plot_mus[tuple(tract)]=[k,w]\n",
    "                        \n",
    "                        for mutation in tract:\n",
    "                            site=int(mutation[1:-1])\n",
    "                            positions.add(site)\n",
    "\n",
    "unique_xs={x: i for i,x in enumerate(sorted(positions))}\n",
    "\n",
    "bound_colours=['k']\n",
    "\n",
    "for t,tract in enumerate(sorted(plot_mus,key=lambda t: min([int(m[1:-1]) for m in t]))): ## iterate over identified tracts\n",
    "    \n",
    "    bc=bound_colours[t%len(bound_colours)]\n",
    "    \n",
    "    k,w=plot_mus[tract] ## get the two branches involved\n",
    "    \n",
    "    A=[k.name] if k.branchType=='leaf' else list(k.leaves) ## get tip name(s) on branch\n",
    "    B=[w.name] if w.branchType=='leaf' else list(w.leaves)\n",
    "    \n",
    "    get_site=lambda mutation: int(mutation[1:-1])\n",
    "    sites=[unique_xs[x] for x in map(get_site,tract)] ## identify relative positions of mutated sites\n",
    "    x1,x2=min(sites),max(sites) ## get tract edges\n",
    "    \n",
    "    A_span=[y_coords[y] for y in A] ## get y positions of tip(s)\n",
    "    B_span=[y_coords[y] for y in B]\n",
    "\n",
    "    A_bounds=Rectangle((x1-0.5,min(A_span)-0.5),(x2-x1)+1,(max(A_span)-min(A_span)+1),facecolor='none',edgecolor=bc,ls='-',lw=2,zorder=100) ## bounding box for tract A\n",
    "    \n",
    "    B_bounds=Rectangle((x1-0.5,min(B_span)-0.5),(x2-x1)+1,(max(B_span)-min(B_span)+1),facecolor='none',edgecolor=bc,ls='-',lw=2,zorder=100) ## bounding box for tract B\n",
    "    \n",
    "    ax2.add_patch(A_bounds)\n",
    "    ax2.add_patch(B_bounds)\n",
    "    \n",
    "    y1,y2=min([max(A_span)+0.5,max(B_span)+0.5]),max([min(A_span)-0.5,min(B_span)-0.5]) ## upper and lower edges of the two bounding boxes\n",
    "    \n",
    "    ax2.add_patch(Rectangle((x1-0.5,y1),x2-x1+1,y2-y1,facecolor=bc,edgecolor='none',zorder=0,alpha=0.1))\n",
    "    \n",
    "    for mutation in tract: ## iterate over every mutation in tract\n",
    "        x=int(mutation[1:-1]) ## get site\n",
    "        \n",
    "        for tip in A+B: ## iterate over tips involved\n",
    "            y=y_coords[tip] ## get tip's y position\n",
    "            m=ll.getExternal(lambda q: q.name==tip)[0].traits['seq'][x] ## get the state of tip at site in question\n",
    "            \n",
    "            ax2.add_patch(Rectangle((unique_xs[x]-0.5,y-0.5),1,1,facecolor=colour[m],edgecolor='none',zorder=2)) ## add nucleotide rectangle in correct colour\n",
    "            ax2.text(unique_xs[x],y,m,ha='center',va='center',zorder=10000) ## add label too\n",
    "    \n",
    "ax2.set_xticks(sorted(list(unique_xs.values())))\n",
    "ax2.set_xticklabels(sorted(list(unique_xs.keys())),size=24,rotation=90)\n",
    "\n",
    "ax2.plot()\n",
    "\n",
    "[ax.spines[loc].set_visible(False) for loc in ax.spines if loc not in ['bottom']]\n",
    "[ax2.spines[loc].set_visible(False) for loc in ax2.spines if loc not in ['bottom']]\n",
    "ax.tick_params(axis='y',size=0,labelsize=0)\n",
    "ax2.tick_params(axis='y',size=0,labelsize=0)\n",
    "\n",
    "ax2.set_xlabel('relative position',size=40)\n",
    "\n",
    "ax2.hlines(np.arange(0.5,ll.ySpan+0.5,1),-0.5,len(unique_xs),ls='--',color='lightgrey')\n",
    "ax2.set_xlim(-0.5,len(unique_xs)-0.5)\n",
    "ax.set_ylim(0,ll.ySpan)\n",
    "ax2.set_ylim(ax.get_ylim())\n",
    "\n",
    "ax3 = fig.add_axes([0.17, 0.87, 0.76, 0.05], frame_on=False)\n",
    "ax3.plot([0,1],[0,0],ls='-',color='grey',lw=8)\n",
    "ax3.text(-0.003,0,'1',size=40,ha='right',va='center')\n",
    "ax3.text(1+0.003,0,'%d'%(len(ll.root.traits['seq'])),size=40,ha='left',va='center')\n",
    "ax3.scatter([x/max(unique_xs.keys()) for x in unique_xs.keys()],[0 for y in unique_xs],s=300,marker='|',lw=2,facecolor='k',edgecolor='none',zorder=10)\n",
    "ax3.set_yticks([])\n",
    "ax3.tick_params(size=0,labelsize=0)\n",
    "ax3.set_xlim(-0.085,1.085)\n",
    "\n",
    "plt.savefig(os.path.join(base_path, 'figures', 'figure_S5A.pdf'),dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T09:06:09.130380Z",
     "start_time": "2020-10-20T09:06:04.111262Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "full_tree=bt.loadNewick(pathlib.Path(base_path).joinpath('data', 'figure_1','Anellos.RAxML.newick').as_posix())\n",
    "full_tree.treeStats()\n",
    "\n",
    "all_tips=[k.name.split('|')[1] for k in ll.getExternal()]\n",
    "\n",
    "fig = plt.figure(figsize=(10, 30),facecolor='w') \n",
    "gs = gridspec.GridSpec(1,1,wspace=0.01) \n",
    "\n",
    "ax=plt.subplot(gs[0],facecolor='w')\n",
    "\n",
    "target=lambda k: k.branchType=='leaf' and (k.name.split('|')[1] in all_tips or k.name.split('|')[1]=='NODE_1_length_3730_cov_142.561361')\n",
    "\n",
    "for k in ll.getExternal():\n",
    "    if k.name.split('|')[1] not in [w.name.split('|')[1] for w in full_tree.getExternal()]:\n",
    "        print(k.name)\n",
    "        \n",
    "full_tree.plotPoints(ax,target=target,colour='indianred',size=100,zorder=1000)\n",
    "\n",
    "print(len(full_tree.getBranches(target)),len(all_tips))\n",
    "\n",
    "root_path=set()\n",
    "for k in full_tree.getBranches(target):\n",
    "    cur_node=k\n",
    "    while cur_node:\n",
    "        root_path.add(cur_node)\n",
    "        cur_node=cur_node.parent\n",
    "\n",
    "target=lambda k: k in root_path\n",
    "        \n",
    "full_tree.plotTree(ax,connection_type='elbow',target=target,colour='k',zorder=100)\n",
    "full_tree.plotTree(ax,connection_type='elbow',target=lambda w: target(w)==False,colour='lightgrey',zorder=1)\n",
    "\n",
    "[ax.spines[loc].set_visible(False) for loc in ax.spines if loc not in ['bottom']]\n",
    "ax.set_yticklabels([])\n",
    "ax.set_yticks([])\n",
    "ax.set_ylim(-5,full_tree.ySpan+5)\n",
    "ax.set_xlim(-0.01*full_tree.treeHeight,full_tree.treeHeight*1.01)\n",
    "\n",
    "plt.savefig(os.path.join(base_path, 'figures', 'figure_S5B.pdf'),dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T09:06:15.681095Z",
     "start_time": "2020-10-20T09:06:09.139251Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "cfml_path='ClonalFrameML'\n",
    "\n",
    "stem='5UTR'\n",
    "aln_path=pathlib.Path(data_path).joinpath('%s.fasta'%(stem)).as_posix()\n",
    "tree_path=pathlib.Path(data_path).joinpath('%s.newick'%(stem)).as_posix()\n",
    "kappa=2.3\n",
    "out_stem=stem\n",
    "bt_utils.reconstruct_ancestral_states(cfml_path,aln_path,tree_path,kappa,out_stem)\n",
    "\n",
    "ll=bt.loadNewick(tree_path)\n",
    "ll.treeStats()\n",
    "\n",
    "bt_utils.annotate_tree(ll,tree_path.replace('.newick','.labelled_tree.newick'),aln_path.replace('.fasta','.reconstructed.fasta'))\n",
    "\n",
    "\n",
    "full_tree=bt.loadNewick(pathlib.Path(base_path).joinpath('data', 'figure_1','Anellos.RAxML.newick').as_posix())\n",
    "full_tree=full_tree.collapseBranches(lambda k: k.length<1/100)\n",
    "full_tree.treeStats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-10-20T09:14:07.985472Z",
     "start_time": "2020-10-20T09:06:15.683471Z"
    }
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10, 20),facecolor='w') \n",
    "gs = gridspec.GridSpec(1,2,width_ratios=[4,10],wspace=0.04) \n",
    "\n",
    "ax=plt.subplot(gs[0],facecolor='w')\n",
    "ax2=plt.subplot(gs[1],facecolor='w')\n",
    "\n",
    "seqs={}\n",
    "for seq in SeqIO.parse(aln_path,'fasta'):\n",
    "    seqs[seq.id]=seq.seq\n",
    "    \n",
    "alnL=len(seqs[seq.id])\n",
    "polymorphic=[i for i in range(alnL) if len(set([s[i] for s in seqs.values()]))>1]\n",
    "print(len(polymorphic),alnL)\n",
    "\n",
    "genus=lambda k: k.name.split('|')[-1].split('_')[0]\n",
    "genus_colours={'Alpha': 'indianred','Beta': 'steelblue', 'Gamma': '#906AA4'}\n",
    "\n",
    "colour={'A': 'indianred', 'C': 'skyblue', 'G': 'yellow', 'T': 'seagreen', '-': 'grey', 'N': 'grey'}\n",
    "\n",
    "for k in ll.getExternal():\n",
    "    \n",
    "    ax.add_patch(Rectangle((k.x,k.y-0.5),1+ll.treeHeight-k.x,1,facecolor=genus_colours[genus(k)],edgecolor='none',zorder=0))\n",
    "    \n",
    "    for i in range(alnL):\n",
    "        ax2.add_patch(Rectangle((i-0.5+1,k.y-0.5),1,1,facecolor=colour[seqs[k.name][i]],edgecolor='none',zorder=1))\n",
    "\n",
    "ll.plotTree(ax,zorder=5)\n",
    "\n",
    "colour=lambda k: genus_colours[genus(k)]\n",
    "\n",
    "ax.xaxis.set_major_locator(mpl.ticker.MultipleLocator(0.3))\n",
    "\n",
    "ax.tick_params(labelsize=18)\n",
    "ax.set_xlabel('divergence (nt subs/site)',size=20)\n",
    "ax2.set_xlabel('alignment position',size=20)\n",
    "\n",
    "ax2.set_xlim(-0.5,alnL+0.5)\n",
    "ax.set_xlim(-0.01*ll.treeHeight,1.05*ll.treeHeight)\n",
    "\n",
    "ax.set_ylim(-2,ll.ySpan+2)\n",
    "ax2.set_ylim(ax.get_ylim())\n",
    "\n",
    "ax.set_yticks([])\n",
    "ax.set_yticklabels([])\n",
    "ax2.set_yticks([])\n",
    "ax2.set_yticklabels([])\n",
    "\n",
    "[ax.spines[loc].set_visible(False) for loc in ax.spines if loc not in ['bottom']]\n",
    "[ax2.spines[loc].set_visible(False) for loc in ax2.spines if loc not in ['bottom']]\n",
    "\n",
    "plt.savefig(os.path.join(base_path, 'figures', 'figure_S5C.pdf'),dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "6145e1abf69061c5329082bc7ec51bd3f5f08c08c17674746f4f23a8991aef83"
  },
  "kernelspec": {
   "display_name": "Python 3.8.2 64-bit ('paper_gytis': conda)",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}